sklearn/preprocessing/_encoders.py

```diff
@@ -622,6 +623,14 @@ class OrdinalEncoder(_BaseEncoder):
     dtype : number type, default np.float64
         Desired dtype of output.
 
+    unknown_value : 'error' or int, default='error'
```

Comment:
should this be called `handle_unknown` for consistency with the OneHotEncoder?

Reply:
Hm, not sure. We used another name to emphasize that we can choose the encoded value here. But you are right, in principle, it is a similar thing.

Reply:
cc @thomasjpfan @jnothman

Reply:
It's slightly strange to have `handle_unknown=-10`, since `-10` is not an "action".

Being the same name may case some confusion, since a user may set it to `'ignore'`, (which should error). Or the inverse may happen: if `handle_unknown` is allowed to be an integer here, a user may set it to be an integer in `OneHotEncoder`.

Those two reasons was why I suggested a different name for this PR.

Reply:
> a user may set it to 'ignore', (which should error). Or the inverse may happen: if handle_unknown is allowed to be an integer here, a user may set it to be an integer in OneHotEncoder

For me that's not really an issue: users would get an error immediately, it's not like the OHE is going to have an unexpected behavior.
We already have other places where the same parameter name accepts slightly different values depending on the nature of the estimator.


This parameter is about how unknown values are handled in transform, so its semantic is  exactly the same as OHE's `handle_unknown`. I think it should be called the same.

Also, at least to me, the current `unknown_value` name suggests that the parameter means "this is what an unknown value looks like in fit", as in e.g. "treat these values as missing", which is a different thing.

> It's slightly strange to have handle_unknown=-10, since -10 is not an "action".

An alternative is to have `handle_unkown='replace'` with an additional `replace_value` parameter.

Maybe others should weigh in, consistency is important.

Reply:
> An alternative is to have handle_unkown='replace' with an additional replace_value parameter.

If we were to go with `handle_unknown` I would prefer this, since the semantics of `handle_unknown` would be clear. Although adding two parameters for this one feature seems a little much.

In https://github.com/scikit-learn/scikit-learn/pull/16959#issuecomment-619435557 I suggested `unknown_encoded_value`, which I thought would be more clear.

Yea lets see what other think.

Reply:
Any further opinions?

Reply:
ping @scikit-learn/core-devs for additional inputs.

Let's also discuss it during the next dev meeting (end of the month) if no consensus is reached

Reply:
> This parameter is about how unknown values are handled in transform, so its semantic is exactly the same as OHE's handle_unknown. I think it should be called the same.

My personal view is to favor having `handle_unknown` as in OHE, having it accept the same values as in OHE, which, as mentioned above requires adding another argument. That last argument could take a default value "out_of_range", which would automatically find an out-of-range value.

Reply:
I am fine to change it that way. Would you agree, @thomasjpfan ?

Reply:
I think 'unknown_value' is fine instead of having two parameters.

Reply:
What happens when one day we decide to have a 'ignore' option to handling missing values in the OrdinalEncoder? Or any other strategy than purely replacing unknown values?

We'd have to introduce a new parameter at that point, and that new parameter would have to be mutually exclusive with `unknown_value`. That would not be ideal in terms of usage, and that would be inconsistent with the OHE.

Also, if one day we want to support NaNs in the OrdinalEncoder and treat them in a specific way, then `unknown_value` will become an ambiguous parameter name.

Finally, just regarding the name: when I see an "unknown_value" parameter name, I intuitively assume this parameter tells me how to recognize unknown values, not how to treat them.

By using `unknown_value` as a parameter, I am worried that we are preventing ourselves from cleanly and consistently supporting alternatives strategies in the future.

Reply:
I wonder if instead of providing an encoded integer value to map the unknown value, one should not all the user to provide the name of an existing or new category name:

```
encoder = OrdinalEncoder(unknown_category="other")
```

If the "other" category exists in the training set, then any unknown value passed at prediction time will be mapped to the same integer code. If not, then a new integer code will be created for this new 'garbage' category.

This has the added benefit that we can then `inverse_transform` encoded unknowns.

Reply:
I am not sure way the default value for `unknown_category` should be. I cannot be a string such as "error" or "raise" as it would be ambiguous with a category name.

It could be `None` (to mean "raise and error") but then what it we want to use `None` as a missing value marker and make it possible to pass missing values in `OrdinalEncoder` in the future... It would be nice to be able to map unknown values to the missing value encoded value and have `inverse_transform` work.

We could have a combination of 2 options:

```
OrdinalEncoder(handle_unknown="use_category", unknown_category="other")
```

which would make it possible to also implement:

```
OneHotEncoder(handle_unknown="use_category", unknown_category="other")
```

to get a consistent behavior across encoder. But in that case the possible values for `OrdinalEncoder.handle_unknown` would be `{"error", "use_category"}` while for `OneHotEncoder.handle_unknown` it would be `{"error", "ignore", "use_category"}``.

That is, the "ignore" would be meaningless for `OrdinalEncoder`.

Reply:
Using a combo of 2 hyperparam is a bit verbose but at least it's explicit and it will make it possible to keep a flexible API to add missing value support in the future.

Reply:
I would be in favor of having `handle_unknown` with an additional parameter. In terms of names, I would prefer the following:

`OrdinalEncoder(handle_unknown="replace", replace_value="other")`


I agree "ignore" doesn't make much sense for `OrdinalEncoder` (which is totally fine IMO).

Reply:
I am now in favor of the two parameter option:

```py
OrdinalEncoder(handle_unknown="use_category", unknown_category="other")
```

with the behavior suggested by @ogrisel 

> If the "other" category exists in the training set, then any unknown value passed at prediction time will be mapped to the same integer code. If not, then a new integer code will be created for this new 'garbage' category.

Is everyone else okay with this?

Reply:
Sorry but I still think that `unknown_category` is not an intuitive name https://github.com/scikit-learn/scikit-learn/pull/16959#discussion_r430475624 and it can conflict with future plans.

> If the "other" category exists in the training set

I think the second parameter (`unknown_category` or `replace_value`) should specify an *encoded* value, not a raw value. Otherwise this will conflict when raw values are integers.

Reply:
> Otherwise this will conflict when raw values are integers.

If the `unknown_category` is seen in prediction time, unknowns will be mapped to it. If not, it will be placed into `categories_` and ordered lexically. The downside is that this would not work when `unknown_category='other'` and numerical input.

The benefit of specifying the encoded value is that it allows algorithms afterwards to recognize what is "unknown" and have special handling of it. 

The other approach is to not have an option and just encode it as `-2` similar to [category_encoders](http://contrib.scikit-learn.org/category_encoders/ordinal.html) does. Missing values would map to `-1`. (I am choosing these encoding scheme to match pandas a little more closely.)

```py
OrdinalEncoder(handle_unknown="use_value")
```

and in the future:

```py
OrdinalEncoder(handle_missing="use_value")
```

I think this would do what most users want with `OridinalEncoder`.

Reply:
Sorry but I don't follow anymore.

I think `unknown_category`/`replace_value` should specify an encoded value, i.e. an int. If it happens that this int is also used to map a known category, then that's fine. We could have an 'auto' option that would map it to `n_known_categories`.

I don't understand how this parameter could potentially be used to specify a raw category value.

Reply:
> I don't understand how this parameter could potentially be used to specify a raw category value.

For that option (which I am +0.25 now on):

```py
X = np.array([['a', 'b', 'd']], dtype=object).T
enc = OridinalEncoder(handle_unknown='use_category',
				      unknown_category='c').fit(X)
enc.transform([['c'], ['f'], ['d'])
# unknown category 'f' is mapped to 2 (raw value 'c')
# array([[2, 2, 3]])
```

I think your suggestion is

```py
X = np.array([['a', 'b', 'd']], dtype=object).T
enc = OridinalEncoder(handle_unknown='use_category',
				      unknown_category=-1).fit(X)
enc.transform([['c'], ['f'], ['d'])
# array([[-1, -1, 2]])
```

Based on how much discussion we have on this, I am moving toward not having the `unknown_category` at first and always map unknown to `-2`. This way we do not have a second parameter. We can always add this option in another PR.

Reply:
```py
X = np.array([['a', 'b', 'd']], dtype=object).T
enc = OridinalEncoder(handle_unknown='use_category',
				      unknown_category='c').fit(X)
```

As I mentioned before this can only be supported if we assume that X has no integer values. If X has integer values then there is now way to know whether the parameter refers to raw values or encoded values. So I don't think we should consider this option.

>  map unknown to -2. 

Why -2?

Reply:
> Why -2?

There is no particular reason, other than I want `-1` to be missing. If we have a demand for enabling mapping to other integers, we can add the parameter then. I think this solves 95% of the problem already with unknown values in `OrdinalEncoder`.

Reply:
> I think the second parameter (unknown_category or replace_value) should specify an encoded value, not a raw value. Otherwise this will conflict when raw values are integers.

I still like the ability to be able to pass a category (in input category space) for the unknown values since it then make it possible to implement `inverse_transform` on test data unknown values which would not be possible when we only pass the integer value.

We could support the two options though.

- `OrdinalEncoder(handle_unknown="use_category", unknown_category="unknown")`
- `OrdinalEncoder(handle_unknown="use_value", unkown_encoded_value=-1)`


Or even just use the `handle_unknown="use_category"` option but make it possible to specify both an input space category and an encoded space value:

- `OrdinalEncoder(handle_unknown="use_category", unknown_category="other", unkown_encoded_value=-1)`

But then we need to be careful if `unknown_category` is already part of the labels of the training set.

Another motivation for making it possible to pass the input space category is to make it easy to map unknown values to a meaningful garbage category that is already present in the training set (e.g. "other", "misc", "background").

Reply:
What about having two parameters, one to decide on the handling (use category or encoded value for unknowns) and one on the value for unknowns (with meaning depending on the handling decision):
- `OrdinalEncoder(handle_unknown="use_category", unknown="other")`

- `OrdinalEncoder(handle_unknown="use_encoded_value", unknown=-1)`
Not sure if we should use a default for category usage. For encoded value usage I like either -1 or -999.

Reply:
I find my original [suggestion](https://github.com/scikit-learn/scikit-learn/pull/16959/files#r432309229) more explicit but I am also fine with yours if others like it better. For the default encoded value I would rather leave it to `None` and force the user to be explicit.

> Not sure if we should use a default for category usage. For encoded value usage I like either -1 or -999.

The default values would be:
```python
`OrdinalEncoder(handle_unknown="error", unknown_value=None)`
```
which would not raise any error at fit time but would raise an error at `predict` time if unknown values are encountered.

```python
`OrdinalEncoder(handle_unknown="use_category", unknown_value=None)`
```

```python
`OrdinalEncoder(handle_unknown="use_encoded_value", unknown_value=None)`
```

would raise an error at fit time asking the user to ask the user  to set `unknown_value` to a value with a valid dtype at fit time (e.g. an integer of `handle_unknown == "use_encoded_value"` or a `str` if `handle_unknown="use_category"` and all category values in the training set as `str` instances.